rfcs/error-identifiers.md

```diff
@@ -0,0 +1,172 @@
+to give per-module identifiers instead of trying to pick global numbers and
+avoiding conflicts between two errors trying to use the same number.)
+
+
+## Impact on the compiler implementation
+
+An important question with this proposal is the maintainability impact on the compilr codebase. We sketched an implementation to see how invasive it would be.
+
+Currently our error-printing logic has code looking as follows, in each module reporting new errors:
+
+```ocaml
+let report_error ~loc env = function
+  | Constructor_arity_mismatch(lid, expected, provided) ->
+      Location.errorf ~loc
+       "@[The constructor %a@ expects %i argument(s),@ \
+        but is applied here to %i argument(s)@]"
+       longident lid expected provided
+  | [..]
+```
+
+We assume a `identifier_of_error : error -> Location.error_identifier` in each such module,
+so the change to `report_error` would be minimal:
+
+```ocaml
+let report_error ~loc env err =
+  let identifier = identifier_of_error err in
+  match err with
+  | Constructor_arity_mismatch(lid, expected, provided) ->
+      Location.errorf ~loc ~identifier
+       "@[The constructor %a@ expects %i argument(s),@ \
+        but is applied here to %i argument(s)@]"
+       longident lid expected provided
+  | [..]
+```
+
+### Assigning identifiers to error values
+
+This cannot easily be done automatically, because the error number
+must be stable, even if the error type changes (a constructor is
+renamed, reordered, split in two, several constructors are factored
+into one...).
+
+We thus propose to have a manually-written `identifier_of_error` function
+
+```ocaml
+let identifier_of_error err =
+  let group = "tyco" in
+  let open Location in
+  match err with
+  | Constructor_arity_mismatch _ -> { group; num = 7 }
+  | [...]
+```
+
+(Note: in general there is no one-to-one mapping between constructors
+in the error type and the user-level concept of "errors", two
+constructors could be the same error or one constructor could be
+different errors depending on its argument.)
+
+The exhaustivity check for pattern-matching on this function
+guarantees that we always assign am identifier to all errors, and in
+particular don't forget picking an identifier for new errors.
+
+
+#### Advanced
+
+If we wanted to be able to generate nice documentation for each error,
+or at least compute statically the set of error identifiers currently in
+use, we could attach an "example" of each error (an OCaml value
+belonging to this error case, chosen to be representative for this
+error and read nicely when printed by the error report) to each case
+of the `identifier_of_error` function using a bespoke attribute
+`[error.example]`:
+
+```ocaml
+let identifier_of_error err =
+  let group = "tyco" in
+  let open Location in
+  match[@error.table] err with
+  | Constructor_arity_mismatch _ ->
+    [@error.example (Constructor_arity_mismatch (Lident "foo", 1, 2))]
+    { group; num = 7 }
```

Comment:
Do we want to move the attribute on the `num` constant?
```ocaml
{ group;
  num = 7 [@error.example(Constructor_arity_mismatch (Lident "foo", 1, 2))];
}
```
that would make the extraction slightly easier.

Reply:
I moved it below the record literal, which is more correct than my previous "floating attribute" sketch and not much more difficult to support (but easier on the eye) than your proposal.